core/src/main/scala/akka/kafka/internal/ExternalSingleSourceLogic.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.kafka.internal
+
+import akka.actor.{ActorRef, Terminated}
+import akka.kafka.KafkaConsumerActor.RequestMessages
+import akka.kafka.Subscription.{Assignment, AssignmentWithOffset, TopicSubscription, TopicSubscriptionPattern}
+import akka.kafka.{KafkaConsumerActor, Subscription}
+import akka.stream.stage.GraphStageLogic.StageActor
+import akka.stream.stage.{GraphStageLogic, OutHandler}
+import akka.stream.{ActorMaterializer, SourceShape}
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.tailrec
+
+private[kafka] abstract class ExternalSingleSourceLogic[K, V, Msg](
+    val shape: SourceShape[Msg],
+    consumer: ActorRef,
+    subscription: Subscription
+) extends GraphStageLogic(shape) with PromiseControl with MessageBuilder[K, V, Msg] {
+  var tps = Set.empty[TopicPartition]
+  var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+  var requested = false
+  var self: StageActor = _
+  implicit lazy val ec = ActorMaterializer.downcast(materializer).executionContext
+
+  override def preStart(): Unit = {
+    super.preStart()
+    setKeepGoing(true)
+    subscription match {
+      case TopicSubscription(topics) =>
+        consumer ! KafkaConsumerActor.Subscribe(topics, KafkaConsumerActor.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
```

Comment:
We do not care about sender in this case. It is just a setup for KafkaConsumer. `sender` used only in `RequestMessages`.